lib/Context/index.ts

```diff
@@ -1,170 +1,38 @@
-  api: DA;
-  handlers?: Handler[];
-  services?: Record<string, any>;
-}
-
-export interface State {
-  turn?: StorageState;
-  stack: FrameState[];
-  storage: StorageState;
-  variables: StorageState;
-}
-
-export enum Action {
-  IDLE,
-  RUNNING,
-  END,
-}
-
-class Context<DA extends DataAPI = DataAPI> extends AbstractLifecycle {
-  public turn: Store;
-
-  public stack: Stack;
-
-  // storage variables
-  public storage: Store;
-
-  // global variables
-  public variables: Store;
-
-  public trace: Trace;
-
-  // services
-  public services: Record<string, any>;
-
-  public api: DA;
-
-  private action: Action = Action.IDLE;
-
-  private handlers: Handler[];
-
-  private programManager: ProgramManager;
-
-  constructor(
-    public versionID: string,
-    state: State,
-    private request: Request | null = null,
-    { services = {}, handlers = [], api }: Options<DA>,
-    events: Lifecycle
-  ) {
-    super(events);
-
-    const createEvent = <K extends EventType>(type: K) => (event: Event<K>) => this.callEvent(type, event);
-
-    this.services = services;
-    this.handlers = handlers;
-    this.api = api;
-
-    this.stack = new Stack(state.stack, {
-      willChange: createEvent(EventType.stackWillChange),
-      didChange: createEvent(EventType.stackDidChange),
-    });
-
-    this.turn = new Store(state.turn, {
-      didUpdate: createEvent(EventType.turnDidUpdate),
-      willUpdate: createEvent(EventType.turnWillUpdate),
-    });
-
-    this.storage = new Store(state.storage, {
-      didUpdate: createEvent(EventType.storageDidUpdate),
-      willUpdate: createEvent(EventType.storageWillUpdate),
-    });
-
-    this.variables = new Store(state.variables, {
-      didUpdate: createEvent(EventType.variablesDidUpdate),
-      willUpdate: createEvent(EventType.variablesWillUpdate),
-    });
-
-    this.trace = new Trace(this);
+export class ContextBuilder<R> {
+  private pipes: ContextHandler<R>[][] = [];
 
-    this.programManager = new ProgramManager(this);
+  addHandlers(...handlers: ContextHandler<R>[]) {
+    this.pipes.push(handlers);
+    return this;
   }
 
-  getRequest(): Request | null {
-    return this.request;
-  }
-
-  public setAction(type: Action): void {
-    this.action = type;
-  }
-
-  public getAction(): Action {
-    return this.action;
-  }
-
-  public end(): void {
-    this.setAction(Action.END);
-  }
-
-  public hasEnded(): boolean {
-    return this.getAction() === Action.END;
-  }
-
-  public async callEvent<K extends EventType>(type: K, event: Event<K>) {
-    await super.callEvent<K>(type, event, this);
-  }
+  async handle(_request: Context<R>) {
+    let request: Context<R> = _request;
 
-  public getProgram(programID: string) {
-    return this.programManager.get(programID);
-  }
-
-  public async update(): Promise<void> {
-    try {
-      await this.callEvent(EventType.updateWillExecute, {});
+    for (const handlers of this.pipes) {
+      for (const handler of handlers) {
```

Comment:
I guess we should set the `end` flag to false before running the pipe, the reason is below. And I guess it is better to rename `end` into `pipeEnd` or something similar.

```suggestion
    for (const handlers of this.pipes) {
      request.end = false;
    
      for (const handler of handlers) {
```